keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,518 @@
+ - a feature flag shall control whether new loadbalancer Services with mixed protcol configuration can be created or not
+ - we must add a note to the documentation that if such Service is created then it may break things after a rollback - it depends on the cloud provider implementation
+ - the update of such Services shall be possible even if the feature flag is OFF. This is to prepare for the next release when the feature flag is removed from the create path, too, and after a rollback to the first release the update of existing Service objects must be possible
+ - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. Either via supporting such Service definitions, or clearly indicating to the users that the Service could not be processed as specified. As we can see from our analysis some CPIs support other protocols than TCP and UDP in the Service definitions, while others support only TCP and UDP. That is the term "mixed protocol support" does not always mean that all possible protocol values are supported by a CPI. For this reason a nicely behaving CPI shall 
+ - indicate clearly to the user what ports with what protocols have been opened on the LB
+ - preferably not create any Cloud LB resources if the Service definition contains unsupported protocols.
+ 
+ In order to provide a way for the CPI to indicate port statuses to the user we would add the following new `portStatus` list of port statuses to `Service.status.loadBalancer.ingress`, so the CPI can indicate the status of the LB. Also we add a `conditions` list of conditions to the `Service.status.loadBalancer`, with the first official condition `LoadBalancerMixedProtocolNotSupported` defined.
+
+```json
+"io.k8s.api.core.v1.ServiceCondition": {
+  "description": "ServiceCondition contains details for the current condition of this Service.",
+  "properties": {
+    "type": {
+      "description": "Type is the type of the condition. Known conditions are \"LoadBalancerMixedProtocolNotSupported\".\n\nThe  \"LoadBalancerMixedProtocolNotSupported\" condition with \"True\" status means that the cloud provider implementation could not create the requested load-balancer with the specified set of ports because that set contains different protocol values for the ports, and such a configuration is not supported either by the cloud provider or by the load balancer or both.",
+      "type": "string"
+    },
+    "status": {
+      "description": "Status is the status of the condition. Can be True, False, Unknown.",
+      "type": "string"
+    },
+    "message": {
+      "description": "Human-readable message indicating details about last transition.",
+      "type": "string"
+    },
+    "reason": {
+      "description": "Unique, one-word, CamelCase reason for the condition's last transition.",
+      "type": "string"
+    }
+  },
+  "required": [
+    "type",
+    "status"
+  ],
+  "type": "object"
+},
+"io.k8s.api.core.v1.PortStatus": {
```

Comment:
Is kube-proxy expected to consume this in any way into iptables/IPVS rules or is this just for reporting? Might be worth clarifying this

Reply:
Good question.  If IPVS and iptables set up captures for the LBVIP + proto + port, and the LB itself doesn't support mixed protocols, should kube-proxy also not capture them?

Reply:
I wonder about the exact handling. I mean, no rules at all, or rather drop those packets that are anyway not handled by the LB on the external interface?

Reply:
So, I added a kube-proxy chapter. Please check and if you are OK with that please remove the hold flag from this one.